feat(service-worker): change default SW strategy behavior

sonukapoor opened the pull request:
Previously, there was a chance that the SW will never register
when there is a long-running task or recurring timeout. This commit
fixes this, by first trying to wait for the app to stabilize, however,
if that does not happen, then the SW will register after 30 seconds.

Closes #34464

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #34464


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

sonukapoor commented:
@alxhub / @gkalpak My local lint and tests are all passing - not sure why CI is complaining. Can you help with this?

mary-poppins commented:
You can preview 21e583d at https://pr35870-21e583d.ngbuilds.io/.
You can preview 0749d35 at https://pr35870-0749d35.ngbuilds.io/.

gkalpak commented:
Thx, @sonukapoor, for taking this on. Can you please remove the formatting changes? They make the PR hard to review and the formatter is unhappy (you are probably running the wrong commands locally if `lint` passes).

Also, how are you running the tests locally? The failures on CI seem expected.

sonukapoor commented:
@gkalpak. I am running yarn lint and yarn test in the aio folder.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -105,9 +105,11 @@ export function ngswAppInitializer(
-          const appRef = injector.get<ApplicationRef>(ApplicationRef);
-          readyToRegister$ = appRef.isStable.pipe(filter(stable => stable));
-          break;
+          // Our default is when the app is stable, or after 30 seconds
```
This is not the place to document the default behavior. This branch is about `registerWhenStable`. (The fact that we use `registerWhenStable` as the default value is done in a [different part of the code][1].)

[1]: https://github.com/angular/angular/blob/f714a54f018277c1113cbac11f5c4735e9e08da6/packages/service-worker/src/module.ts#L99

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -105,9 +105,11 @@ export function ngswAppInitializer(
+          // Our default is when the app is stable, or after 30 seconds
+          // This ensures that the SW is still registered, in the case of a long running 
+          // or recurring timeout. 
+          readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0] || 30000)));
```
Thinking about it again, I don't think we should change the behavior of `registerWhenStable` (in case it is used explicitly by someone), only the default value. It should also be possible for people to wait for stability indefinitely (if they so wish).

Therefore, I suggest the following changes:
- When not passing an argument (e.g. `'registerWhenStable'`), keep the current behavior (i.e. wait for stability indefinitely).
- When passing a timeout argument (e.g. `'registerWhenStable:10000'`), wait for stability or for the timeout to expire.
- Change the [default strategy][1] from `'registerWhenStable'` to `'registerWhenStable:30000'` (timeout up for debate).

[1]: https://github.com/angular/angular/blob/f714a54f018277c1113cbac11f5c4735e9e08da6/packages/service-worker/src/module.ts#L99

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -105,9 +105,11 @@ export function ngswAppInitializer(
+          // Our default is when the app is stable, or after 30 seconds
+          // This ensures that the SW is still registered, in the case of a long running 
+          // or recurring timeout. 
+          readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0] || 30000)));
```
Hm...I just realized: Does this `delay` prevent the app from stabilizing? If that's the case (which it probably is), we should run it outside the `NgZone`. Same for other `registerWithDelay` strategy.

sonukapoor replied:
@gkalpak Good catch 👍

gkalpak reviewed (commented):
Here are some more comments from a preliminary high-level review.

gkalpak commented:
> @gkalpak. I am running yarn lint and yarn test in the aio folder. When I try to run similar commands within root then I get all kind of strange errors from bazel

These changes are outside the `aio/` directory, so you need to run commands at the root of the repo :grin:
Can you open a separate issue with the errors you are getting from bazel?

sonukapoor commented:
> Can you open a separate issue with the errors you are getting from bazel?

@gkalpak, Please see here: #35891

mary-poppins commented:
You can preview 5d6531f at https://pr35870-5d6531f.ngbuilds.io/.

mary-poppins commented:
You can preview 4b8b57a at https://pr35870-4b8b57a.ngbuilds.io/.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -7,8 +7,8 @@
 import {isPlatformBrowser} from '@angular/common';
-import {APP_INITIALIZER, ApplicationRef, InjectionToken, Injector, ModuleWithProviders, NgModule, PLATFORM_ID} from '@angular/core';
-import {Observable, of } from 'rxjs';
+import {APP_INITIALIZER, ApplicationRef, InjectionToken, Injector, ModuleWithProviders, NgModule, PLATFORM_ID, NgZone} from '@angular/core';
```
Please order imported symbols.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
+      // Our default is when the app is stable, or after 30 seconds
+      // This ensures that the SW is still registered, in the case of a long running 
+      // or recurring timeout. 
+      const [strategy, ...args] = (options.registrationStrategy || 'registerWhenStable:30000').split(':');
```
Move this change to a separate commit.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
-          readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          ngZone.runOutsideAngular(() => {
+            readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          });
```
Have you verified that:
a. This prevents the app from stabilizng without `runOutsideangular`?
b. The app can stabilize with `runOutsideAngular` here (vs where the observable is subscribed to)?

We also need some test(s) to verify the above.

sonukapoor replied:
I was going to test all of this using unit tests but since I can’t run bazel I am stuck on it.

gkalpak replied:
It turns out that we need to `runOutsideAngular()` when subscribing to the observable (because that is where the timeouts are created). Creating the observable (but not subscribing to it) outside the zone does nothing to allow the app to stabilize :grin:

I will move the `runOutsideAngular()` call to where we subscribe.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
+              } else {
+                readyToRegister$ = whenStable(injector);
+              }
+            });
```
```suggestion
          ngZone.runOutsideAngular(() => {
            if (+args[0]) {
              readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0])));
            } else {
              readyToRegister$ = whenStable(injector);
            }
          });
```

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -75,7 +75,8 @@ export const SCRIPT = new InjectionToken<string>('NGSW_REGISTER_SCRIPT');
     injector: Injector, script: string, options: SwRegistrationOptions,
-    platformId: string): Function {
+    platformId: string, 
+    ngZone: NgZone): Function {
```
In order to avoid unnecessary work, I suggest we use the `injector` to get `NgZone` only when/if needed.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
-          const appRef = injector.get<ApplicationRef>(ApplicationRef);
-          readyToRegister$ = appRef.isStable.pipe(filter(stable => stable));
+            ngZone.runOutsideAngular(() => {
+              if (+args[0]) {
```
This will treat 0 the same as not providing an argument. I think this is not correct. Explicitly specifying `0` as an argument should be respected.

sonukapoor replied:
Good point

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
+              } else {
+                readyToRegister$ = whenStable(injector);
+              }
+            });
```
Or rather:

```suggestion
          readyToRegister$ = (args[0] === undefined) ?
              whenStable(injector) :
              ngZone.runOutsideAngular(() => merge(whenStable(injector), of(null).pipe(delay(+args[0]))));
```

But, again, we should verify whether `runOutsideAngular()` is needed here or when subscribing to the observable.

sonukapoor replied:
@gkalpak, you are right. The app will never stabilize. Here is the documentation:

https://angular.io/api/core/ApplicationRef#isstable-examples-and-caveats

> the application will never be stable if you start any kind of recurrent asynchronous task when the application starts (for example for a polling process, started with a setInterval, a setTimeout or using RxJS operators like interval);

By wrapping the code block in the `runOutsideAngular` function, we should be fine.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -117,13 +128,18 @@ export function ngswAppInitializer(
     // Don't return anything to avoid blocking the application until the SW is registered.
     // Catch and log the error if SW registration fails to avoid uncaught rejection warning.
-    readyToRegister$.pipe(take(1)).subscribe(
+    readyToRegister$!.pipe(take(1)).subscribe(
```
This should not be needed with my suggested refactoring.

sonukapoor replied:
I tried that, but typescript is still complaining about the error `Variable 'readyToRegister$' is used before being assigned.ts(2454)`

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
           break;
         case 'registerWithDelay':
-          readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          ngZone.runOutsideAngular(() => {
```
Please move this change to a separate commit.

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
-          readyToRegister$ = appRef.isStable.pipe(filter(stable => stable));
+            ngZone.runOutsideAngular(() => {
+              if (+args[0]) {
+                readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0])));
```
Might be worth creating a helper for these `of(null).pipe(delay(+args[0]))`.

gkalpak commented on packages/service-worker/test/module_spec.ts:
```diff
@@ -177,6 +177,21 @@ describe('ServiceWorkerModule', () => {
            expect(swRegisterSpy).toHaveBeenCalledWith('sw.js', {scope: undefined});
          }));
 
+        it('registers the SW when the app doesnt stabilizes with `registerWhenStable`, but uses the timeout', fakeAsync(() => {
```
```suggestion
        it('registers the SW when the app does not stabilize after `<timeout>` with `registerWhenStable:<timeout>`', fakeAsync(() => {
```

gkalpak commented:
BTW, @sonukapoor, you don't need to run all the tests when you only update a specific package. For this PR for example, you could run `yarn bazel test //packages/service-worker/...` or `yarn bazel test //packages/service-worker/test`.
(Replace `bazel` with `ibazel` in the above commands to run in "watch  mode".)

sonukapoor commented:
Thanks. I will do that once I get bazel working. Until then it’s very difficult to run any tests. I will make the code changes for now and test everything once the bazel issue is resolved.

sonukapoor commented:
@gkalpak, I think I have addressed all of your comments. However, I am not able to run any tests, because of the previously mentioned errors with bazel. I tried the single tests but get the same errors. Let me know if the code/commits are otherwise fine.

mary-poppins commented:
You can preview 92c17cc at https://pr35870-92c17cc.ngbuilds.io/.

mary-poppins commented:
You can preview d939416 at https://pr35870-d939416.ngbuilds.io/.

sonukapoor commented:
@gkalpak, something is wrong with CI, my PWA score locally with the same arguments is at passing:

![image](https://user-images.githubusercontent.com/59734/76405967-5d810180-635f-11ea-951f-7e89a9f642cd.png)

mary-poppins commented:
You can preview 41e1b73 at https://pr35870-41e1b73.ngbuilds.io/.

mary-poppins commented:
You can preview b0f127c at https://pr35870-b0f127c.ngbuilds.io/.

mary-poppins commented:
You can preview 0e13be2 at https://pr35870-0e13be2.ngbuilds.io/.

mary-poppins commented:
You can preview 487d3e6 at https://pr35870-487d3e6.ngbuilds.io/.

mary-poppins commented:
You can preview 2eacc62 at https://pr35870-2eacc62.ngbuilds.io/.

mary-poppins commented:
You can preview afcdf51 at https://pr35870-afcdf51.ngbuilds.io/.

sonukapoor commented:
@gkalpak I borrowed a mac for this PR. It seems that the tests are failing because of the timeout that we have set on the service worker. It does not complete in time with the timeout set in protractor/jasmine. It seems to be set to `10secs`. I have tried to set that but couldn't figure it out. I tried setting it directly in the `it`s third function argument or using `jasmine.DEFAULT_TIMEOUT_INTERVAL = 31000`, but nothing worked. I then simply reduced the time in the service worker to 8 secs instead of 30. Now the tests seem to pass fine - see [here](https://app.circleci.com/pipelines/github/angular/angular/11854/workflows/dd226ce1-5290-41d1-b628-d15e6084491a/jobs/658504). However, the lighthouse audits are failing now. They are passing for me on windows and mac. I am not sure why `CI` is complaining about this.

gkalpak commented:
@googlebot I consent.

sonukapoor committed:
feat(service-worker): support timeout in `registerWhenStable` SW registration strategy

Previously, when using the `registerWhenStable` ServiceWorker
registration strategy (which is also the default) Angular would wait
indefinitely for the [app to stabilize][1], before registering the
ServiceWorker script. This could lead to a situation where the
ServiceWorker would never be registered when there was a long-running
task (such as an interval or recurring timeout).

Such tasks can often be started by a 3rd-party dependency (beyond the
developer's control or even without them realizing). In addition, this
situation is particularly hard to detect, because the ServiceWorker is
typically not used during development and on production builds a
previous ServiceWorker instance might be already active.

This commit enhances the `registerWhenStable` registration strategy by
adding support for an optional `<timeout>` argument, which guarantees
that the ServiceWorker will be registered when the timeout expires, even
if the app has not stabilized yet.

For example, with `registerWhenStable:5000` the ServiceWorker will be
registered as soon as the app stabilizes or after 5 seconds if the app
has not stabilized by then.

Related to #34464.

[1]: https://angular.io/api/core/ApplicationRef#is-stable-examples

gkalpak committed:
fixup! feat(service-worker): support timeout in `registerWhenStable` SW registration strategy

sonukapoor committed:
fix(service-worker): by default register the SW after 30s even the app never stabilizes

Previously, when using the default ServiceWorker registration strategy
Angular would wait indefinitely for the [app to stabilize][1], before
registering the ServiceWorker script. This could lead to a situation
where the ServiceWorker would never be registered when there was a
long-running task (such as an interval or recurring timeout).

Such tasks can often be started by a 3rd-party dependency (beyond the
developer's control or even without them realizing). In addition, this
situation is particularly hard to detect, because the ServiceWorker is
typically not used during development and on production builds a
previous ServiceWorker instance might be already active.

This commit fixes this by changing the default registration strategy
from `registerWhenStable` to `registerWhenStable:30000`, which will
ensure that the ServiceWorker will be registered after 30s at the
latest, even if the app has not stabilized by then.

Fixes #34464

gkalpak committed:
fixup! fix(service-worker): by default register the SW after 30s even the app never stabilizes

sonukapoor committed:
fix(service-worker): prevent SW registration strategies from affecting app stabilization

Previously, some of the built-in ServiceWorker registration strategies,
namely `registerWithDelay:<timeout>` and `registerWhenStable:<timeout>`,
would register potentially long-running timeout, thus preventing the app
from stabilizing before the timeouts expired. This was especially
problematic for the `registerWhenStable:<timeout>` strategy, which waits
for the app to stabilize, because the strategy itself would prevent the
app from stabilizing and thus the ServiceWorker would always be
registered after the timeout.

This commit fixes this by subscribing to the registration strategy
observable outside the Angular zone, thus not affecting the app's
stabilization.

gkalpak committed:
fixup! fix(service-worker): prevent SW registration strategies from affecting app stabilization

gkalpak commented:
Rebased on master, updated commit messages and fixed up commits. Let's see what CI thinks :smiley:
@sonukapoor, please take a look and make sure I didn't miss anything 🙏

sonukapoor commented:
> Rebased on master, updated commit messages and fixed up commits. Let's see what CI thinks 😃
> @sonukapoor, please take a look and make sure I didn't miss anything 🙏

Looks good to me. Thanks for the updates - very much appreciated.

mary-poppins commented:
You can preview 5bf2fe2 at https://pr35870-5bf2fe2.ngbuilds.io/.

gkalpak reviewed (approved):
Thx, @sonukapoor ✨

mary-poppins commented:
You can preview 22bf4cf at https://pr35870-22bf4cf.ngbuilds.io/.
